### PR TITLE
This adds support for detaching the hessian contribution to `A_grad` in the sparse solvers

### PR DESCRIPTION
**Pull request by luisenp**
See #425 for more context.

### CODE REVIEW

**Review by bamos — APPROVED** [review]

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py [hunk 1]
```diff
@@ -577,16 +573,16 @@ def _step(
...
         tensor_dict, err = self._compute_retracted_tensors_and_error(
             delta, converged_indices, force_update
         )
-        if truncated_grad_loop:
-            # For "implicit" or "truncated", the grad-attached steps are just GN steps
-            # So, we need to avoid calling `_complete_step`, as it's likely to reject
-            # the step computed
+        if delta_forced_to_gn:
+            # If delta has been forced to be a GN step (for implicit diff),
+            # the we need to make sure we ignore any reject indices computed by
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:584]: typo: the -> then

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py [hunk 2]
```diff
@@ -362,7 +360,7 @@ def _optimize_loop(
                     info.status[:] = NonlinearOptimizerStatus.FAIL
                     return iters_done
 
-            if truncated_grad_loop:
+            if last_implicit_diff_step:
                 # This is a "secret" option that is currently being tested in the
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:364]: `last_implicit_diff_step` -> `_last_implicit_diff_step` ?

**Review by mhmukadam — APPROVED** [review]

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py (hunk 2 above)
Comment (luisenp) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py]: Changed.

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py (hunk 1 above)
Comment (luisenp) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py]: Fixed.

**Pull request merged**